In [5]:
!dropdb -U student project2

dropdb: error: database removal failed: ERROR:  database "project2" is being accessed by other users
DETAIL:  There is 1 other session using the database.


In [2]:
!createdb -U student project2

In [2]:
%load_ext sql

In [3]:
%sql postgresql://student@/project2

In [5]:
!pwd

/home/ubuntu


## Question 1 
Please see the attached document where we put the ERDiagram and relational schema.

## Question 2 Create tables

+ convert all txt files into csv files. 

In [23]:
!csvformat -d "|" weball20.txt > all_candidates.csv

In [ ]:
!csvformat -d "|" cn.txt > candidate_master.csv

In [ ]:
!csvformat -d "|" cm.txt > committee_master.csv

In [6]:
!csvformat -d "|" ccl.txt > linkage.csv

+ Extract the specific columns that we need 

In [ ]:
!csvcut -c 1,6,7,8,9,10,11,18,28,29,30 -H all_candidates.csv> All_candidates.csv 

In [ ]:
!csvcut -c1,2,3,4,5,6,7,8,9,10,11,12,13,14,15 candidate_master.csv -H > Candidate_master.csv

In [ ]:
!csvcut -c1,2,3,4,5,6,7,8,9,10,11,12,13,14,15 committee_master.csv -H > Committee_master.csv

In [ ]:
!csvcut -c1,4,7 -H linkage.csv > Linkage.csv

+ take a look at the data, and check on the NULL constraints of data. 

In [30]:
!csvstat all_candidates.csv

/home/ubuntu/.local/lib/python3.8/site-packages/agate/utils.py:292: DuplicateColumnWarning: Column name "0" already exists in Table. Column will be renamed to "0_2".
/home/ubuntu/.local/lib/python3.8/site-packages/agate/utils.py:292: DuplicateColumnWarning: Column name "0" already exists in Table. Column will be renamed to "0_3".
/home/ubuntu/.local/lib/python3.8/site-packages/agate/utils.py:292: DuplicateColumnWarning: Column name "367.52" already exists in Table. Column will be renamed to "367.52_2".
/home/ubuntu/.local/lib/python3.8/site-packages/agate/utils.py:292: DuplicateColumnWarning: Column name "0" already exists in Table. Column will be renamed to "0_4".
/home/ubuntu/.local/lib/python3.8/site-packages/agate/utils.py:292: DuplicateColumnWarning: Column name "0" already exists in Table. Column will be renamed to "0_5".
/home/ubuntu/.local/lib/python3.8/site-packages/agate/utils.py:292: DuplicateColumnWarning: Column name "0" already exists in Table. Column will be renamed to "

#### create All_candidates table

In [7]:
%%sql
DROP TABLE IF EXISTS All_candidates Cascade;

CREATE TABLE All_candidates
(
  CAND_ID VARCHAR(9) NOT NULL,
  TTL_RECEIPTS NUMERIC(14,2) NOT NULL,
  TRANS_FROM_AUTH NUMERIC(14,2) NOT NULL,
  TTL_DISB NUMERIC(14,2) NOT NULL,
  TRANS_TO_AUTH NUMERIC(14,2) NOT NULL,
  COH_BOP NUMERIC(14,2) NOT NULL,
  COH_COP NUMERIC(14,2) NOT NULL,
  TTL_INDIV_CONTRIB NUMERIC(14,2) NOT NULL,
  CVG_END_DT DATE NOT NULL,
  INDIV_REFUNDS NUMERIC(14,2) NOT NULL,
  CMTE_REFUNDS NUMERIC(14,2) NOT NULL,
  PRIMARY KEY (CAND_ID)
);

 * postgresql://student@/project2
Done.
Done.


[]

+ From the result of "csvstat", we can see that there are no null value in all attributes in All_candidates file. 
+ Therefore, we add not null for all attributes in All_candidates table. 

+ Check the attributes in All_candidates

In [ ]:
%%sql
SELECT *
FROM All_candidates;

#### Create Candidate_master table

In [ ]:
!csvstat Candidate_master.csv

In [9]:
%%sql
DROP TABLE IF EXISTS Candidate_master cascade;

CREATE TABLE Candidate_master
(
  CAND_ID VARCHAR(9) NOT NULL,
  CAND_NAME VARCHAR(200) NOT NULL,
  CAND_PTY_AFFILIATION VARCHAR(3),
  CAND_ELECTION_YR NUMERIC(4) NOT NULL,
  CAND_OFFICE_ST VARCHAR(2) NOT NULL,
  CAND_OFFICE VARCHAR(1) NOT NULL,
  CAND_OFFICE_DISTRICT VARCHAR(2),
  CAND_ICI VARCHAR(8),
  CAND_STATUS VARCHAR(1) NOT NULL,
  CAND_PCC VARCHAR(9),
  CAND_ST1 VARCHAR(34),
  CAND_ST2 VARCHAR(34),
  CAND_CITY VARCHAR(30),
  CAND_ST VARCHAR(2),
  CAND_ZIP VARCHAR(9),
  PRIMARY KEY (CAND_ID)
);

 * postgresql://student@/project2
Done.
Done.


[]

+ Check the attributes in Candidate_master

In [10]:
%%sql
SELECT *
FROM Candidate_master;

 * postgresql://student@/project2
0 rows affected.


[]

#### Create Committee_master table

+ Take a look at the data. 

In [ ]:
!csvstat Committee_master.csv

In [23]:
%%sql
DROP TABLE IF EXISTS Committee_master cascade;
CREATE TABLE Committee_master
(
  CMTE_ID VARCHAR(9) NOT NULL,
  CMTE_NM VARCHAR(200),
  TRES_NM VARCHAR(90),
  CMTE_ST1 VARCHAR(34),
  CMTE_ST2 VARCHAR(34),
  CMTE_CITY VARCHAR(30),
  CMTE_ST VARCHAR(2),
  CMTE_ZIP VARCHAR(9),
  CMTE_DSGN VARCHAR(1),
  CMTE_TP VARCHAR(1),
  CMTE_PTY_AFFILIATION VARCHAR(3),
  CMTE_FILING_FREQ VARCHAR(1) NOT NULL,
  ORG_TP VARCHAR(1),
  CONNECTED_ORG_NM VARCHAR(200),
  CAND_ID VARCHAR(9),
  PRIMARY KEY (CMTE_ID)
);


 * postgresql://student@/project2
Done.
Done.


[]

In [24]:
%%sql
SELECT *
FROM Committee_master;

 * postgresql://student@/project2
0 rows affected.


[]

In [ ]:
!csvstat Linkage.csv

In [22]:
%%sql
DROP TABLE IF EXISTS Linkage cascade;
CREATE TABLE Linkage
(
  CAND_ID VARCHAR(9) NOT NULL,
  CMTE_ID VARCHAR(9) NOT NULL,
  LINKAGE_ID NUMERIC(12) NOT NULL
--FOREIGN KEY (CAND_ID) REFERENCES Candidate_master(CAND_ID),
--FOREIGN KEY (CMTE_ID) REFERENCES Committee_master(CMTE_ID)
);

 * postgresql://student@/project2
Done.
Done.


[]

In [14]:
%%sql
SELECT *
FROM Linkage;

 * postgresql://student@/project2
0 rows affected.


[]

In [15]:
%%sql
DROP TABLE IF EXISTS Donate cascade;
CREATE TABLE Donate
(
  SUB_ID NUMERIC(19) NOT NULL,
  CMTE_ID VARCHAR(9) NOT NULL
--FOREIGN KEY (SUB_ID) REFERENCES Contribution_by_individuals(SUB_ID),
--FOREIGN KEY (CMTE_ID) REFERENCES Committee_master(CMTE_ID)
);

 * postgresql://student@/project2
Done.
Done.


[]

In [16]:
%%sql
SELECT *
FROM Donate;

 * postgresql://student@/project2
0 rows affected.


[]

#### Create contribution table

In [13]:
%%sql
DROP TABLE IF EXISTS Contribution_by_individuals cascade;
CREATE TABLE Contribution_by_individuals
(
    CMTE_ID varchar(9) NOT NULL,
    AMNDT_IND varchar(1) check (AMNDT_IND in('N','A','T')),
    RPT_TP varchar(3),
    TRANSACTION_PGI varchar(5),
    IMAGE_NUM varchar(18),
    TRANSACTION_TP varchar(3),
    ENTITY_TP varchar(3),
    NAME varchar(200),
    CITY varchar(30),
    STATE varchar(2),
    ZIP_CODE varchar(9),
    EMPLOYER varchar(38),
    OCCUPATION varchar(38),
    TRANSACTION_DT varchar(8),
    TRANSACTION_AMT numeric(14,2),
    OTHER_ID varchar(9),
    TRAN_ID varchar(32) NOT NULL,
    FILE_NUM varchar(22) NOT NULL,
    MEMO_CD varchar(1),
    MEMO_TEXT varchar(100),
    SUB_ID numeric(19) NOT NULL,
    UNIQUE(SUB_ID)
)

 * postgresql://student@/project2
Done.
Done.


[]

In [ ]:
# Use the csvstack command to stack the 13 contribution by individual data files for 2020 and form a single csv file.
!csvstack -d "|" itcont_2020_20*_2020*.txt > itcon.csv

## Question 3

+ Using csvcut to check whether the data exists duplicates. 

In [ ]:
!csvcut -c1,2 All_candidates.csv | sort -n | uniq -c | sort -rn | head -5

In [17]:
%%sql
COPY All_candidates FROM '/home/ubuntu/All_candidates.csv'
CSV
HEADER;

 * postgresql://student@/project2
3883 rows affected.


[]

+ take a look at the data set.

In [ ]:
%%sql
SELECT *
FROM All_candidates
limit 5;

+ Use csvcut to check any duplicated rows in the column corresponding to the primary key.
+ From the result showed that there are no duplicated rows in All_candidates file.

In [ ]:
!csvcut -c1,2 Candidate_master.csv | sort -n | uniq -c | sort -rn | head -5

+ From the result of above, we can see that there duplicates exist in Candidate_master.csv that we should delete these rows and roload. 

In [18]:
%%sql
COPY Candidate_master FROM '/home/ubuntu/Candidate_master.csv'
CSV
HEADER;

 * postgresql://student@/project2
7500 rows affected.


[]

In [19]:
%%sql
SELECT *
FROM Candidate_master
limit 5;

 * postgresql://student@/project2
5 rows affected.


[('H0AK00105', 'LAMB, THOMAS', 'NNE', Decimal('2020'), 'AK', 'H', '0', 'C', 'N', 'C00607515', '1861 W LAKE LUCILLE DR', None, 'WASILLA', 'AK', '99654'),
 ('H0AK00113', 'TUGATUK, RAY SEAN', 'DEM', Decimal('2020'), 'AK', 'H', '0', 'C', 'N', None, 'PO BOX 172', None, 'MANAKOTAK', 'AK', '99628'),
 ('H0AK01046', 'CATALANO, THOMAS', 'OTH', Decimal('2020'), 'AK', 'H', '0', None, 'N', None, '188 WEST NORTHERN LIGHTS BOULEVARD', None, 'ANCHORAGE', 'AK', '99503'),
 ('H0AL01055', 'CARL, JERRY LEE, JR', 'REP', Decimal('2020'), 'AL', 'H', '1', 'O', 'C', 'C00697789', 'PO BOX 852138', None, 'MOBILE', 'AL', '36685'),
 ('H0AL01063', 'LAMBERT, DOUGLAS WESTLEY III', 'REP', Decimal('2020'), 'AL', 'H', '1', 'O', 'C', 'C00701557', '7194 STILLWATER BLVD', None, 'SPANISH FORT', 'AL', '36527')]

+ Use select from except to find the CAND_IDs from All_candidates that are not present on the candidate_master.

In [20]:
%%sql
select cand_id
from All_candidates
except 
select cand_id 
from Candidate_master;

 * postgresql://student@/project2
10 rows affected.


[('P80003916',),
 ('P00006890',),
 ('S6NC00324',),
 ('S0CA00447',),
 ('P20003851',),
 ('P60006244',),
 ('P00006197',),
 ('S8MD00310',),
 ('P00015651',),
 ('H8NC00023',)]

+ Based on the result of last cell, there are existing 10 rows in All_candidates that Candidate_master does not include.
+ That is why we can not use cand_id as the foregin key to connect both tables. To solve this problem, we need to remove these 10 CAND_ID from the all_candidate dataset or add these 10 CAND_ID into the candidate_master dataset.

In [ ]:
!csvcut -c1,2 Committee_master.csv | sort -n | uniq -c | sort -rn | head -5

+ From the result of above, we can see that there duplicates exist in Candidate_master.csv that we should delete these duplicate rows and roload. 

In [25]:
%%sql
COPY Committee_master FROM '/home/ubuntu/Committee_master.csv'
CSV
HEADER;

 * postgresql://student@/project2
17615 rows affected.


[]

In [27]:
%%sql
SELECT *
FROM Committee_master
limit 5;

 * postgresql://student@/project2
5 rows affected.


[('C00000059', 'HALLMARK CARDS PAC', 'SARAH MOE', '2501 MCGEE', 'MD #500', 'KANSAS CITY', 'MO', '64108', 'U', 'Q', 'UNK', 'M', 'C', None, None),
 ('C00000422', 'AMERICAN MEDICAL ASSOCIATION POLITICAL ACTION COMMITTEE', 'WALKER, KEVIN MR.', '25 MASSACHUSETTS AVE, NW', 'SUITE 600', 'WASHINGTON', 'DC', '200017400', 'B', 'Q', None, 'M', 'M', 'DELAWARE MEDICAL PAC', None),
 ('C00000489', 'D R I V E POLITICAL FUND CHAPTER 886', 'JERRY SIMS JR', '3528 W RENO', None, 'OKLAHOMA CITY', 'OK', '73107', 'U', 'N', None, 'Q', 'L', None, None),
 ('C00000547', 'KANSAS MEDICAL SOCIETY POLITICAL ACTION COMMITTEE', 'JERRY SLAUGHTER', '623 SW 10TH AVE', None, 'TOPEKA', 'KS', '666121627', 'U', 'Q', 'UNK', 'Q', 'M', 'KANSAS MEDICAL SOCIETY', None),
 ('C00000638', 'INDIANA STATE MEDICAL ASSOCIATION POLITICAL ACTION COMMITTEE', 'ACHENBACH, GRANT MR.', '322 CANAL WALK, CANAL LEVEL', None, 'INDIANAPOLIS', 'IN', '46202', 'U', 'Q', None, 'T', 'M', None, None)]

In [12]:
!csvcut -c1,4,7 -H linkage.csv > Linkage.csv

In [13]:
!csvcut -c1,2,3 Linkage.csv | sort -n | uniq -c | sort -rn | head -5

      1 a,d,g
      1 S8WY00197,C00671776,224671
      1 S8WY00189,C00663146,223873
      1 S8WV00143,C00710889,228546
      1 S8WV00143,C00689471,225486
sort: write failed: 'standard output': Broken pipe
sort: write error


In [28]:
%%sql
COPY Linkage FROM '/home/ubuntu/Linkage.csv'
CSV
HEADER;

 * postgresql://student@/project2
6829 rows affected.


[]

In [15]:
%%sql
SELECT *
FROM Linkage
limit 5;

 * postgresql://student@/project2
5 rows affected.


[('C00713602', 'C00712851', Decimal('228963')),
 ('H0AK00105', 'C00607515', Decimal('229250')),
 ('H0AL01055', 'C00697789', Decimal('226125')),
 ('H0AL01063', 'C00701557', Decimal('227053')),
 ('H0AL01071', 'C00701409', Decimal('227054'))]

In [13]:
%%sql
COPY Contribution_by_individuals FROM '/home/ubuntu/itcon.csv'
CSV
HEADER;

 * postgresql://student@/project2
(psycopg2.errors.DiskFull) could not extend file "base/16799/16833.4": wrote only 4096 of 8192 bytes at block 577976
HINT:  Check free disk space.
CONTEXT:  COPY contribution_by_individuals, line 20455901

[SQL: COPY Contribution_by_individuals FROM '/home/ubuntu/itcon.csv'
CSV
HEADER;]
(Background on this error at: http://sqlalche.me/e/e3q8)


+ Comment: due to the limit disk space that we have, I can only copy 20455901 reconds into the Contribution_by_individual table. 

In [ ]:
!csvstat --count contribution_by_individual.csv

In [29]:
%%sql
select CAND_ID
from Linkage
except
select CAND_ID
from Candidate_master;

 * postgresql://student@/project2
80 rows affected.


[('P80003916',),
 ('P20005443',),
 ('P00014084',),
 ('P00012195',),
 ('P00015990',),
 ('P00007385',),
 ('P00014928',),
 ('P00007781',),
 ('H0TX13202',),
 ('P00006890',),
 ('P00011601',),
 ('P00009001',),
 ('P00013235',),
 ('P00010215',),
 ('P00009522',),
 ('P20000055',),
 ('P00009118',),
 ('S6NC00324',),
 ('P00015701',),
 ('P00010348',),
 ('P00015966',),
 ('P80003205',),
 ('H0CA41136',),
 ('P00016105',),
 ('P00010249',),
 ('P00008607',),
 ('H0UT01122',),
 ('P00008961',),
 ('P40003873',),
 ('P60006350',),
 ('P00005447',),
 ('S8CT00139',),
 ('P00010777',),
 ('P00011015',),
 ('P00011379',),
 ('P00008383',),
 ('P60022613',),
 ('P00010892',),
 ('P60009156',),
 ('P00015545',),
 ('P00013334',),
 ('P00012534',),
 ('P00008821',),
 ('P80006133',),
 ('P20005435',),
 ('P20003851',),
 ('P00007658',),
 ('P00014563',),
 ('P00013060',),
 ('P00009811',),
 ('P00014167',),
 ('P00015172',),
 ('P00012203',),
 ('P80003239',),
 ('P00014373',),
 ('P00012286',),
 ('P00010009',),
 ('H8FL14209',),
 ('P00015255',),
 ('P60006244',),
 ('P60009610',),
 ('P60005774',),
 ('P00006197',),
 ('P00011775',),
 ('P00012948',),
 ('H8TX26108',),
 ('P00016089',),
 ('S8MD00310',),
 ('P00015651',),
 ('P60021441',),
 ('P00015263',),
 ('H8C006211',),
 ('H8PA13158',),
 ('P00012450',),
 ('P00012351',),
 ('P00006791',),
 ('S0CA00462',),
 ('P00012377',),
 ('P00008987',),
 ('C00713602',)]

+ From the result of last cell, we can see there are 80 cand_ids in Linkage but do not exists in the Candidate_master, which means we can not use cand_id as the foregin key to connect both tables. Inorder to solve this problem, we should remove these 80 data from the Linkage dataset or add it into the Candidate_master.

In [30]:
%%sql
select CMTE_ID
from Linkage
except
select CMTE_ID
from Committee_master;

 * postgresql://student@/project2
90 rows affected.


[('C00709600',),
 ('C00692798',),
 ('C00746925',),
 ('C00691816',),
 ('C00727495',),
 ('C00741082',),
 ('C00586024',),
 ('C00666628',),
 ('C00699025',),
 ('C00630004',),
 ('C00731554',),
 ('C00617019',),
 ('C00694760',),
 ('C00731505',),
 ('C00699074',),
 ('C00721316',),
 ('C00652677',),
 ('C00697110',),
 ('C00745364',),
 ('C00676197',),
 ('C00731620',),
 ('C00570192',),
 ('C00742171',),
 ('C00742189',),
 ('C00696674',),
 ('C00702068',),
 ('C00690503',),
 ('C00714162',),
 ('C00730556',),
 ('C00700112',),
 ('C00671487',),
 ('C00690701',),
 ('C00709139',),
 ('C00699355',),
 ('C00699678',),
 ('C00733535',),
 ('C00728188',),
 ('C00735415',),
 ('C00711622',),
 ('C00736082',),
 ('C00746610',),
 ('C00733782',),
 ('C00692574',),
 ('C00679076',),
 ('C00746487',),
 ('C00695536',),
 ('C00696278',),
 ('C00428599',),
 ('C00733493',),
 ('C00694448',),
 ('C00728709',),
 ('C00507947',),
 ('C00680108',),
 ('C00697599',),
 ('C00727438',),
 ('C00727453',),
 ('C00693283',),
 ('C00674184',),
 ('C00734160',),
 ('C00687640',),
 ('C00697177',),
 ('C00710335',),
 ('C00724534',),
 ('C00732313',),
 ('C00646679',),
 ('C00708602',),
 ('C00709246',),
 ('C00585067',),
 ('C00690966',),
 ('C00623553',),
 ('C00694893',),
 ('C00729848',),
 ('C00709931',),
 ('C00742197',),
 ('C00745836',),
 ('C00707885',),
 ('C00715128',),
 ('C00717512',),
 ('C00727446',),
 ('C00693259',),
 ('C00702266',),
 ('C00738344',),
 ('C00688143',),
 ('C00693218',),
 ('C00703884',),
 ('C00624239',),
 ('C00704940',),
 ('C00735340',),
 ('C00700302',),
 ('C00727032',)]

+ From the result of last cell, we can see there are 90 cmte_ids in Linkage but do not exists in the Committee_master, which means we can not use cmte_id as the foregin key to connect both tables. Inorder to solve this problem, we should remove the 90 data from the Linkage dataset or add it into the Committee_master.

+ This is the reason why we can not add foreign key to connect these tables. In order to import the dataset into the tables, we were not connect those tables by foreign key since some datasets are incomplete. Whenever we import the data into the table, it will should an error as there exists unequal data among tables.

### Question 4

In [ ]:
%%sql
SELECT count(*)
FROM All_candidates

In [ ]:
!csvstat --count All_candidates.csv 

In [ ]:
%%sql
SELECT count(*)
FROM Candidate_master

In [ ]:
!csvstat --count Candidate_master.csv 

In [ ]:
%%sql
SELECT count(*)
FROM Committee_master

In [ ]:
!csvstat --count Committee_master.csv

In [ ]:
%%sql
SELECT count(*)
FROM Linkage

In [ ]:
!csvstat --count Linkage.csv

In [ ]:
%%sql
SELECT count(*)
FROM Contribution_by_individual

In [ ]:
!csvstat --count Contribution_by_individual.csv

### Question 5

+ Use the alter statement to add the first name and last name columns to Candidate_master

In [31]:
%%sql
ALTER TABLE Candidate_master
ADD firstname varchar(50),
ADD lastname varchar(50);

 * postgresql://student@/project2
Done.


[]

+ Split firstname and lastname and add them to the newly created colomns

In [32]:
%%sql
UPDATE Candidate_master
SET firstname = split_part(CAND_NAME,',',1),
    lastname = split_part(CAND_NAME,',',2);

 * postgresql://student@/project2
7500 rows affected.


[]

+ Check the results

In [33]:
%%sql
SELECT * 
FROM Candidate_master
limit 5;

 * postgresql://student@/project2
5 rows affected.


[('H2UT00013', 'BUCK, JASON', 'REP', Decimal('2012'), 'UT', 'H', '2', 'C', 'P', 'C00501601', None, None, 'SALT LAKE CITY', None, None, 'BUCK', ' JASON'),
 ('H4TX10028', 'DOGGETT, LLOYD', 'DEM', Decimal('2020'), 'TX', 'H', '35', 'I', 'C', 'C00286500', None, None, 'AUSTIN', None, None, 'DOGGETT', ' LLOYD'),
 ('H6PA01165', 'LEIB, BRYAN', 'DEM', Decimal('2018'), 'PA', 'H', '3', 'C', 'P', 'C00680256', None, None, 'PHILADELPHIA', None, None, 'LEIB', ' BRYAN'),
 ('P00009555', 'MA, JAMES', 'DEM', Decimal('2020'), 'US', 'P', '0', 'C', 'N', None, '20571 CALHOUN AVE', None, 'CALDWELL', 'ID', '83605', 'MA', ' JAMES'),
 ('S0KS00372', 'BERLAND, LANCE', 'REP', Decimal('2020'), 'KS', 'S', '0', 'O', 'N', None, 'PO BOX 161', None, 'CHAPMAN', 'KS', '67431', 'BERLAND', ' LANCE')]

### Question 6

In [22]:
!csvstat -d "," linkage.csv

  1. "C00713602"

	Type of data:          Text
	Contains null values:  False
	Unique values:         6140
	Longest value:         9 characters
	Most common values:    P60022118 (11x)
	                       P80006133 (9x)
	                       S2MT00096 (7x)
	                       S4IA00129 (7x)
	                       P00011361 (6x)

  2. "2019"

	Type of data:          Number
	Contains null values:  False
	Unique values:         38
	Smallest value:        1984
	Largest value:         2929
	Sum:                   13788280
	Mean:                  2019.373169
	Median:                2020
	StDev:                 19.22218
	Most common values:    2020 (4278x)
	                       2018 (1535x)
	                       2016 (278x)
	                       2014 (130x)
	                       2024 (117x)

  3. "2020"

	Type of data:          Number
	Contains null values:  False
	Unique values:         1
	Smallest value:        2020
	Largest value:         2020
	Sum:                   13792

+ Use Committee_master to create the new table "LINKAGE" that mimics linkage file as Committee_master contains both cmte_id and cand_id. 

In [34]:
%%sql
DROP TABLE IF EXISTS LINKAGE cascade;

CREATE TABLE LINKAGE as 
select T1.cand_id, T1.cmte_id 
from Committee_master T1, Committee_master T2 where T1.cand_id = T2.cand_id and T1.cmte_id = T2.cmte_id

 * postgresql://student@/project2
Done.
6520 rows affected.


[]

+ count the number of rows in LINKAGE.

In [35]:
%%sql
select count(*) from LINKAGE;

 * postgresql://student@/project2
1 rows affected.


[(6520,)]

In [36]:
%%sql
select count(CAND_ID) from Linkage;

 * postgresql://student@/project2
1 rows affected.


[(6520,)]

+ Another way to create the linkage that mimics LINKAGE between Candidate_master and Committee_master.

+ Create two new tables with cand_id from Candidate_master and with cmte_id and cand_id from Committee_master and then join the two tables by cand_id to mimic as Linkage table.

In [5]:
%%sql
DROP TABLE IF EXISTS candidate_connect Cascade;
DROP TABLE IF EXISTS  committee_connect Cascade;

create Table candidate_connect (
    CAND_ID      varchar(9) NOT NULL

);
create Table committee_connect (
   CMTE_ID      varchar(9) NOT NULL,
   CAND_ID      varchar(9) DEFAULT NULL
);

 * postgresql://student@/project2
Done.
Done.
Done.
Done.


[]

+ Extract the cand_id data from All_candidates and put into a new file. 
+ Extract the cand_id and cmte_id data from Committee_master and put into a new file

In [37]:
!csvcut -c1 All_candidates.csv > candidate_connect.csv

In [38]:
!csvcut -c1,15 Committee_master.csv  > committee_connect.csv

+ Copy the data into the two new tables: candidate_connect and committee_connect

In [39]:
%%sql
COPY candidate_connect FROM '/home/ubuntu/candidate_connect.csv'
CSV
HEADER;

COPY committee_connect FROM '/home/ubuntu/committee_connect.csv'
CSV
HEADER;

 * postgresql://student@/project2
3883 rows affected.
17615 rows affected.


[]

+ Use UNION to connect two tables via the cand_id,CAND_PCC and cmte_id and generate a new table named newlinkage. 

In [9]:
%%sql
DROP TABLE IF EXISTS newlinkage Cascade;

CREATE TABLE newlinkage as
SELECT cand_id, CAND_PCC from Candidate_master where cand_pcc IS NOT NULL
UNION
SELECT cand_id, cmte_id from committee_connect where cand_id IS NOT NULL

 * postgresql://student@/project2
Done.
6520 rows affected.


[]

In [10]:
%%sql
SELECT count(*) from newlinkage;

 * postgresql://student@/project2
1 rows affected.


[(6520,)]

In [17]:
!csvstat --count linkage.csv

Row count: 6828


+ Create a sequence called link_num and the sequence start from 1 and increment by 1.

In [42]:
%%sql
DROP SEQUENCE  IF EXISTS link_num CASCADE;

CREATE SEQUENCE link_num
START WITH 1 
INCREMENT BY 1;  

 * postgresql://student@/project2
Done.
Done.


[]

+ Add the link_id and insert the link_num into the column.

In [44]:
%%sql
ALTER TABLE newlinkage ADD column link_id numeric(6) NOT NULL DEFAULT Nextval('link_num');

 * postgresql://student@/project2
Done.


[]

In [45]:
%%sql
SELECT * FROM newlinkage 
limit 10

 * postgresql://student@/project2
10 rows affected.


[('H0SC07011', 'C00692053', Decimal('1'), Decimal('1')),
 ('P00005900', 'C00639021', Decimal('2'), Decimal('2')),
 ('H8CA32070', 'C00632885', Decimal('3'), Decimal('3')),
 ('H8AZ02185', 'C00663054', Decimal('4'), Decimal('4')),
 ('H4VA07143', 'C00554949', Decimal('5'), Decimal('5')),
 ('P00012427', 'C00711812', Decimal('6'), Decimal('6')),
 ('H8UT03238', 'C00647339', Decimal('7'), Decimal('7')),
 ('H0AL02160', 'C00723478', Decimal('8'), Decimal('8')),
 ('H8FL06148', 'C00666396', Decimal('9'), Decimal('9')),
 ('H0NY15210', 'C00707281', Decimal('10'), Decimal('10'))]